src/pickup.cpp

```diff
@@ -74,54 +74,98 @@ struct pickup_count {
     int count = 0;
 };
 
+static bool is_valid_auto_pickup( const item *pickup_item )
+{
+    int weight_limit = get_option<int>( "AUTO_PICKUP_WEIGHT_LIMIT" );
+    int volume_limit = get_option<int>( "AUTO_PICKUP_VOL_LIMIT" );
+
+    return pickup_item->volume() > units::from_milliliter( volume_limit * 50 ) ||
+           pickup_item->weight() > weight_limit * 50_gram;
+}
+
+static bool should_auto_pickup( const item *pickup_item )
+{
+    std::string item_name = pickup_item->tname( 1, false );
+    rule_state pickup_state = get_auto_pickup().check_item( item_name );
+
+    if( !is_valid_auto_pickup( pickup_item ) ) {
+        return false;
+    } else if( pickup_state == rule_state::WHITELISTED ) {
+        return true;
+    } else if( pickup_state != rule_state::BLACKLISTED ) {
+        //No prematched pickup rule found, check rules in more detail
+        get_auto_pickup().create_rule( pickup_item );
+
+        if( get_auto_pickup().check_item( item_name ) == rule_state::WHITELISTED ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static std::vector<item_location *> get_pickup_list_from( item_location *container )
```

Comment:
I wouldn't use pointers here. `item_location` already is a kind of pointer to `item`. For the return type just regular item_location and for the parameter reference instead should do what you need.

Reply:
Could you be more specific, are you referring to `get_pickup_list_from` or `should_auto_pickup` function?
Feel free to write a code snippet to help me better understand.

Reply:
I mean `get_pickup_list_from`. The function header would be like this:
```cpp
static std::vector<item_location> get_pickup_list_from( item_location &container )
```

That allows you to simplify the rest of the code a bit by removing some dereference and address-of operations.

Reply:
Would that not result in storing copies of `item_location` in vector? I understand that `item_location` does not store large amounts of data but copying the whole object over would still be slower then copying a pointer. Note that I am still learning C++ so I might not understand the situation but I believe we should optimize where we can.

Reply:
As far as I know that is not an issue for `std::vector` return types, no copying will be done there.

Reply:
That's interesting, do you have some reference where I could read more about vectors and copying?

Reply:
Doing some more in depth reading myself to make sure I'm not telling you wrong stuff, but see `copy elision` and `return value optimization`. That should apply to any object, not just containers like vectors. Basically in most cases just return the object if you don't need the pointer/reference.

Reply:
The folks on Discord say that vectors entries are not exempt from being copied over when emplaced or pushed. However they also say that `item_location` is very light and that the negligible performance gain is not worth the added decrease in readability. So I will change the vectors to gain more code readability, thanks for bringing this to my attention.

Reply:
As of 9098fd443bac136816a3ec1dd459a06990206822 all `item_location` objects should be passed by value instead of reference.